### PR TITLE
[WIP] Wrap scikit-learn

### PR DESCRIPTION
**Pull request by jcrist**
Wrap many models using a few simple solvers:

- Averaged: fit on each chunk, average the results. This works fine for
  most linear models, but isn't guaranteed to be correct (from what I
  understand).
- Chained: for estimators with a ``partial_fit`` method, fits by
  chaining calls to ``partial_fit`` together. Just provides a consistent
  interface to scikit-learn's existing functionality
- Gradient: just for the ``LogisticRegression`` for now, uses the
  previous gradient descent code.

Dask solvers are specified with a ``dask_solver`` keyword to the estimator. Models are wrapped in a declarative way for now, which makes this fairly easy to maintain. Wherever possible existing scikit-learn functionality is used.

Solvers that can return ``Delayed`` objects will do so if ``compute=False`` is passed to the ``fit`` method. If this isn't supported, a nice error is thrown.

### CODE REVIEW

**Comment by mrocklin** [comment]
From this notebook: https://rawgit.com/jcrist/c111c9628de8da345bd3076bdd4f4c8c/raw/f40a1b04e4fde8a7545d09f53e999b02494cce9f/dask_learn_experiments.html

Looks like split-fit-merge is pretty effective for LogisticRegression and SGD.

I'm surprised to see parallelism in the profile plot for SGD

File: dask_learn/solvers.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+              (_intercept_norm(w, c) + step_size * step_len))
+        log('%2d  %.6e %-.2e  %.2e  %.1e' % (k, f, df, dw, step_size))
+        if df < tol:
+            log('Converged')
+            break
+        step_size *= step_growth
+        backtrack_mult = next_backtrack_mult
+    else:
+        log('Maximum Iterations')
+    return w[None, :], np.atleast_1d(c)
```
Comment (ogrisel) [inline dask_learn/solvers.py]: Instead of re-implementing yet-another gradient descent solver, why not re-using the very efficient L-BFGS solver of scipy and use dask to parallelize the gradient computation (`fprime`)?

http://docs.scipy.org/doc/scipy-0.17.1/reference/generated/scipy.optimize.fmin_l_bfgs_b.html

**Comment by ogrisel** [comment]
Also one of the best way to implement parallelized versions of the SGD optimizer is Elastic Averaging SGD as presented in this paper: http://arxiv.org/abs/1412.6651

However I am afraid that this is overkill for fitting linear model that is likely to converge (in validation loss) after just a couple of pass over the training data. 

In my opinion implementing out-of-core efficiently with dask is enough for fitting linear model. A tool that would make it possible to train in parallel a grid of SGD linear models in an out-of-core manner with different values for the regularization and learning rate hyper parameters along with early stopping on a validation is much more practically useful than trying to push parallelism into the SGD solver it-self.

File: dask_learn/solvers.py (hunk 1 above)
Comment (ogrisel) [inline dask_learn/solvers.py]: Note that tensorflow is following this approach in the last comments of this discussion:

https://github.com/tensorflow/tensorflow/issues/446

**Comment by jcrist** [comment]
I'm trying out a second design here: https://github.com/jcrist/dask-learn/pull/2. This doesn't attempt to duplicate classes, but rather provide some control flow primitives for wrapping estimators. Estimators that use dask internally (such as using L-BFGS) can be implemented as separate classes. In this way it's a bit clearer when dask is used to wrap a scikit-learn model, and when the learning algorithm is implemented directly in dask. I'd appreciate some feedback over there.

Thanks for the suggestion of just using scipy.optimize - that's a lot cleaner, and shouldn't be too difficult. I'll give that paper a read as well.
